gpii/node_modules/gpii-oauth2/gpii-oauth2-datastore/src/DbDataStoreUtils.js

```diff
@@ -0,0 +1,408 @@
+/*!
+GPII CouchDB OAuth 2 Data Store
+
+Copyright 2016 OCAD university
+
+Licensed under the New BSD license. You may not use this file except in
+compliance with this License.
+
+The research leading to these results has received funding from the European Union's
+Seventh Framework Programme (FP7/2007-2013) under grant agreement no. 289016.
+
+You may obtain a copy of the License at
+https://github.com/GPII/universal/blob/master/LICENSE.txt
+*/
+
+"use strict";
+
+var fluid = fluid || require("infusion"),
+    gpii = fluid.registerNamespace("gpii"),
+    $ = fluid.registerNamespace("jQuery"),
+    uuid = require("node-uuid");
+
+require("../../gpii-oauth2-utilities/src/OAuth2Utilities.js");
+
+fluid.registerNamespace("gpii.oauth2.dbDataStore");
+
+// All doc types used for saving different records into CouchDB/PouchDB
+gpii.oauth2.dbDataStore.docTypes = {
+    user: "user",
+    gpiiToken: "gpiiToken",
+    client: "client",
+    authDecision: "authDecision",
+    authCode: "authCode",
+    clientCredentialsToken: "clientCredentialsToken"
+};
+
+gpii.oauth2.dbDataStore.findRecord = function (dataSource, termMap, valueNotEmpty, dataProcessFunc) {
+    dataProcessFunc = dataProcessFunc || gpii.oauth2.dbDataStore.CleanUpDoc;
+    var promiseTogo = fluid.promise();
+
+    // Verify required field values are provided instead of undefined
+    var emptyFields = gpii.oauth2.dbDataStore.verifyEmptyFields(termMap, valueNotEmpty);
+
+    if (emptyFields.length > 0) {
+        var error = gpii.oauth2.composeError(gpii.oauth2.errors.missingInput, {fieldName: emptyFields});
+        promiseTogo.reject(error);
+    } else {
+        var promise = dataSource.get(termMap);
+        promise.then(function (data) {
+            // $.isEmptyObject() is to work around the issue when fetching data
+            // using pouch/couch DB views and records are not found, instead of
+            // returning a 404 status code, it returns this object:
+            // { total_rows: 1, offset: 0, rows: [] }
+            // Note the "rows" value is an empty array.
+            // This behavior prevents "kettle.dataSource.CouchDB" -> "notFoundIsEmpty"
```

Comment:
Comment and behaviour is rather unclear. What exactly is it that prevents "notFoundIsEmpty" from returning `undefined`, and do we even want this behaviour? Why are we checking for an empty object rather than a 0-length array? etc.